### PR TITLE
feat: support extra metadata in vectordb

### PR DESCRIPTION
**Pull request by guenhter**
# Pull Request Checklist

### Note to first-time contributors: Please open a discussion post in [Discussions](https://github.com/open-webui/open-webui/discussions) and describe your changes before submitting a pull request.

**Before submitting, make sure you've checked the following:**

- [x] **Target branch:** Please verify that the pull request targets the `dev` branch.
- [x] **Description:** Provide a concise description of the changes made in this pull request.
- [x] **Changelog:** Ensure a changelog entry following the format of [Keep a Changelog](https://keepachangelog.com/) is added at the bottom of the PR description.
- [x] **Documentation:** Have you updated relevant documentation [Open WebUI Docs](https://github.com/open-webui/docs), or other documentation sources?
- [x] **Dependencies:** Are there any new dependencies? Have you updated the dependency versions in the documentation?
- [x] **Testing:** Have you written and run sufficient tests to validate the changes?
- [x] **Code review:** Have you performed a self-review of your code, addressing any coding standard issues and ensuring adherence to the project's coding standards?
- [x] **Prefix:** To clearly categorize this pull request, prefix the pull request title using one of the following:
  - **BREAKING CHANGE**: Significant changes that may affect compatibility
  - **build**: Changes that affect the build system or external dependencies
  - **ci**: Changes to our continuous integration processes or workflows
  - **chore**: Refactor, cleanup, or other non-functional code changes
  - **docs**: Documentation update or addition
  - **feat**: Introduces a new feature or enhancement to the codebase
  - **fix**: Bug fix or error correction
  - **i18n**: Internationalization or localization changes
  - **perf**: Performance improvement
  - **refactor**: Code restructuring for better maintainability, readability, or scalability
  - **style**: Changes that do not affect the meaning of the code (white space, formatting, missing semi-colons, etc.)
  - **test**: Adding missing tests or correcting existing tests
  - **WIP**: Work in progress, a temporary label for incomplete or ongoing work

# Changelog Entry

### Description

This PR is just a preparation for a bigger epic to support more sources for creating knowledge basis, like Sharepoint, Web, ... (https://github.com/open-webui/open-webui/issues/5872, https://github.com/open-webui/open-webui/discussions/13877, https://github.com/open-webui/open-webui/discussions/13713, https://discord.com/channels/1170866489302188073/1241654544639725609/threads/1395395559124238366)

The main aim of this PR is to support storing the extra metadata of table `file` field `meta.data` also in the vector db. Having extra metadata has several use cases:
* allow to store additional information when the file upload is used via API
* extra metadata can be used for better data analysis outside of OWUI
* extra metadata can be used by tools/functions for better customization

Additionally, I've changed how nested meta data is treated. By default nested dict structures were converted so a string (I guess based on the comment this was done because of Chroma). For the Vector DBs which don't support nested structures, nothing changes. The DBs which definitely support nesting in the metadata, nothing is stringified but instead used as is:
* Qdrant
* Pinecone
* Elasticsearch
* OpenSearch

I've tested it with one DB supporting nesting (qdrant) and with one not supporting nesting (chroma)

### Added

- Allow to store extra metadata in vector DB (same meta as in sql table `files` column `meta.data`)

### Changed

- not stringify nested metadata datastructurs for databases which supports nested structures

---

### Additional Information

This also works when the DB is re-indexed, as the metadata is used from the `file` table of the sql db.

This is also contributing as preparation for features like this: https://qdrant.tech/documentation/search-precision/automate-filtering-with-llms/

### Screenshots or Videos

Qdrant point when metadata is nested:
image

### Contributor License Agreement

By submitting this pull request, I confirm that I have read and fully agree to the [Contributor License Agreement (CLA)](/CONTRIBUTOR_LICENSE_AGREEMENT), and I am providing my contributions under its terms.

### CODE REVIEW

**Review by OrangeDoro — COMMENTED** [review]
Hi! I'm a grad student working on a research project about using large language models to automate code review. Based on your commit a921f032a03facd9db20d859c6a78ef7269f9f5f and the changes in backend/open_webui/routers/retrieval.py, my tool generated this comment:
1. **Null Check for `file.meta`**: Ensure that `file` itself is not `None` before accessing its attributes. Consider adding a check for `file` being `None` right after retrieving it with `Files.get_file_by_id(form_data.file_id)`.
2. **Handling of `file.meta`**: Ensure that `file.meta` is always a dictionary or `None`. If it could be of a different type (e.g., a string or a list), this could lead to unexpected behavior.
3. **Data Integrity**: When updating file metadata and content, ensure that there are checks in place to verify that the updates are successful. If an update fails, the system should handle it gracefully to avoid leaving the system in an inconsistent state.
4. **Impact on `save_docs_to_vector_db`**: Verify that the `save_docs_to_vector_db` function can handle the new `"extra"` key appropriately to avoid logical errors or unintended behavior.
5. **Error Handling**: The `try-except` block around the `save_docs_to_vector_db` function does not specify the type of exceptions it is catching. It is advisable to catch specific exceptions to avoid unintended consequences and ensure that security-related issues are not overlooked.
6. **Testing for Edge Cases**: Test scenarios where `file.meta` is `None`, an empty dictionary, or contains various structures to ensure that the logic behaves as expected in all cases.
7. **Handling of `form_data.collection_name`**: Validate `form_data.collection_name` to ensure it is not an empty string or contains invalid characters.
8. **Result Validation**: Check if `result.ids` is not empty before accessing the first element to avoid `IndexError`.
9. **File Path Handling**: Add validation to ensure that `file_path` is a valid path and not an empty string or non-existent path.
10. **Loader Initialization**: Validate all parameters before passing them to the `Loader`, ensuring they are not `None` or invalid.
11. **Dynamic Metadata Handling**: Consider implementing a more flexible metadata handling mechanism, such as a schema or a validation layer that can adapt to different metadata structures.
12. **Magic Strings**: Replace magic strings like `"content"` and `"data"` in the metadata dictionary with constants or enumerations to avoid typos and improve maintainability.
13. **Use of Ternary Operator**: Instead of `add=(True if form_data.collection_name else False)`, it can be simplified to `add=bool(form_data.collection_name)`. This enhances readability.
14. **Avoiding Redundant Code**: Refactor the creation of `Document` instances into a helper function to avoid code duplication.
15. **Consistent Naming Conventions**: Rename `hash` to `file_hash` to clarify its purpose.
16. **Logging Sensitive Information**: The line `log.debug(f"text_content: {text_content}")` could potentially log sensitive information. Ensure that sensitive data is not logged, or implement a mechanism to redact sensitive information before logging.
17. **Testing for New Functionality**: Ensure that there are corresponding tests for the new functionality introduced by the `"extra"` field.

As part of my research, I'm trying to understand how useful these comments are in real-world development. If you have a moment, I'd be super grateful if you could quickly reply to these two yes/no questions:


1. Does this comment provide suggestions from a dimension you hadn’t considered?

2. Do you find this comment helpful?

Thanks a lot for your time and feedback! And sorry again if this message is a bother.

**Comment by guenhter** [comment]
Hi,

I'll take the time to look at you findings:

1. Null Check for file.meta: In theory this can happen but the `file.meta` is used in several other places in this function and the the writer of `file` ensures that `meta` is always set -> I won't change it here even though the finding is right.
2. Handling of file.meta: If we would always check if the type is actually what it should be, then python programs would not be readable any more -> not a good finding
3. Data Integrity: my change is not affecting the way of the update is done -> false finding
4. Impact on save_docs_to_vector_db: the new field is not handled in the code as it is never read anywhere so this is also no relevant finding.
5. Error Handling: my changes are not as critical as they would be secured by any additional try catch. So also no relevant finding.
6. Testing for Edge Cases: I did some testing like empty or even stringified. 
7. Handling of form_data.collection_name: This is outside of the scope of this PR as I don't change anything about the actual writing to the DB so invalid finding
8. Result Validation: same as 7.
9. File Path Handling: This PR is not concerned about the file path at all. Invalid finding
10. Loader Initialization:  same as 7.
11. Dynamic Metadata Handling: not needed.
12. Magic Strings: I don't see a benefit atm
13. Use of Ternary Operator: This is true, but not scope of my PR. The finiding is in the existing code, not in my PR.
14. Avoiding Redundant Code: same as 13.
15. Consistent Naming Conventions: same as 13
16. Logging Sensitive Information:  same as 13
17. Testing for New Functionality: This is very valid, but unfortuantely the python tests are not executable atm. I'm waiting until they are running again and then add some tests regarding this.


All in all, this review comments were costing too much time considering the value gained from it (even though I'm not saying that all of them are bad).

**Comment by tjbck** [comment]
Partially addressed with 6a17ba5b7a74b788e339034b29254ca539ec865e with bug fixes, @guenhter I'm assuming the other vectordb implementation you left out have all been tested?

File: backend/open_webui/retrieval/vector/dbs/milvus.py [hunk 1]
```diff
@@ -311,7 +312,7 @@ def insert(self, collection_name: str, items: list[VectorItem]):
                     "id": item["id"],
                     "vector": item["vector"],
                     "data": {"text": item["text"]},
-                    "metadata": item["metadata"],
+                    "metadata": stringify_nested_datastructures(item["metadata"]),
```
Comment (tjbck) [inline backend/open_webui/retrieval/vector/dbs/milvus.py]: This part would not have worked, see more: 6a17ba5b7a74b788e339034b29254ca539ec865e

**Comment by guenhter** [comment]
> Partially addressed with [6a17ba5](https://github.com/open-webui/open-webui/commit/6a17ba5b7a74b788e339034b29254ca539ec865e) with bug fixes, @guenhter I'm assuming the other vectordb implementation you left out have all been tested?

Hi @tjbck,

because I cannot say to have this tested very well with elastic/opensearch, I've added the `stringify_metadata` in for the moment to be on the save side. The only DB is qdrant I've done extensive testing with deep nesting and thelike.

File: backend/open_webui/retrieval/vector/dbs/opensearch.py [hunk 2]
```diff
@@ -2,6 +2,7 @@
 from opensearchpy.helpers import bulk
 from typing import Optional
 
+from backend.open_webui.retrieval.vector.utils import stringify_metadata
```
Comment (tjbck) [inline backend/open_webui/retrieval/vector/dbs/opensearch.py:5]: Please test your PRs before creating one.

**Comment by tjbck** [comment]
Closing in favour of 7aeca7dee25144e2a3d9776ef7391ece5eb56a87.

**Comment by rgaricano** [comment]
Is it necessary to perform any migration operations on the actual database after applying these commits?

**Comment by guenhter** [comment]
For all databases where the `stringify_metadata` is used, nothing changes. Only for qdrant and pinecone the metadata change but no migration is needed as long as this is metadata is not used. New documents will have proper nested datastructures, old ones have only string values.  If the old documents should also get proper nested structures, a reindex of the vector storage is needed.
